public/js/utils/source-map.js

```diff
@@ -1,160 +1,206 @@
-const { isEnabled, getValue } = require("../feature");
+let sourceMapRequests = new Map();
 
-let sourceMapWorker;
-function restartWorker() {
-  if (sourceMapWorker) {
-    sourceMapWorker.terminate();
-  }
-  sourceMapWorker = new Worker(
-    getValue("baseWorkerURL") + "source-map-worker.js"
-  );
+function clearData() {
+  sourceMapRequests = new Map();
 }
-restartWorker();
 
-function destroy() {
-  if (sourceMapWorker) {
-    sourceMapWorker.terminate();
-    sourceMapWorker = null;
+function _resolveSourceMapURL(source) {
+  if (path.isURL(source.sourceMapURL) || !source.url) {
+    // If it's already a full URL or the source doesn't have a URL,
+    // don't resolve anything.
+    return source.sourceMapURL;
+  } else if (path.isAbsolute(source.sourceMapURL)) {
+    // If it's an absolute path, it should be resolved relative to the
+    // host of the source.
+    const urlObj = URL.parse(source.url);
+    const base = urlObj.protocol + "//" + urlObj.host;
+    return base + source.sourceMapURL;
   }
+  // Otherwise, it's a relative path and should be resolved relative
+  // to the source.
+  return path.dirname(source.url) + "/" + source.sourceMapURL;
 }
 
-const sourceMapTask = function(method) {
-  return function() {
-    const args = Array.prototype.slice.call(arguments);
-    return workerTask(sourceMapWorker, { method, args });
-  };
-};
+/**
+ * Sets the source map's sourceRoot to be relative to the source map url.
+ */
+function _setSourceMapRoot(sourceMap, absSourceMapURL, source) {
+  // No need to do this fiddling if we won't be fetching any sources over the
+  // wire.
+  if (sourceMap.hasContentsOfAllSources()) {
+    return;
+  }
+
+  const base = path.dirname(
+    (absSourceMapURL.indexOf("data:") === 0 && source.url) ?
+      source.url :
+      absSourceMapURL
+  );
 
-const getOriginalSourcePosition = sourceMapTask("getOriginalSourcePosition");
-const getGeneratedSourceLocation = sourceMapTask("getGeneratedSourceLocation");
-const createOriginalSources = sourceMapTask("createOriginalSources");
-const getOriginalSourceUrls = sourceMapTask("getOriginalSourceUrls");
-const getOriginalTexts = sourceMapTask("getOriginalTexts");
-const createSourceMap = sourceMapTask("createSourceMap");
-const clearData = sourceMapTask("clearData");
+  if (sourceMap.sourceRoot) {
+    sourceMap.sourceRoot = path.join(base, sourceMap.sourceRoot);
+  } else {
+    sourceMap.sourceRoot = base;
+  }
 
-function _shouldSourceMap(source) {
-  return isEnabled("sourceMaps") && source.sourceMapURL;
+  return sourceMap;
 }
 
-function isMapped(source) {
-  return _shouldSourceMap(source);
+function originalToGeneratedId(originalId) {
+  const match = originalId.match(/(.*)\/originalSource/);
+  return match ? match[1] : null;
 }
 
-function isOriginal(originalSource) {
-  return !!getGeneratedSourceId(originalSource);
+function generatedToOriginalId(generatedId, url) {
+  return generatedId + "/originalSource-" + md5(url);
```

Comment:
See my notes in the PR description about it. It's not about escaping, just about shortening so the ids don't look crazy long if the URL is long. I'd be fine with just using `url` but I thought we might want it to look better.

We don't use random ids because we want to be able to always go from `(generatedId, url)` to `originalId` consistently. We can call it multiple times and it generates the same `originalId`. This allows functions to avoid having to dig into the redux state just to get the id.